src/ocamlc_loc/ocamlc_loc.ml

```diff
@@ -0,0 +1,204 @@
+module Re = Dune_re
+module Dyn = Stdune.Dyn
+
+type severity =
+  | Error
+  | Warning of
+      { code : int
+      ; name : string
+      }
+
+type message =
+  | Raw of string
+  | Structured of
+      { file_excerpt : string option
+      ; message : string
+      ; severity : severity
+      }
+
+type loc =
+  { path : string
+  ; line : [ `Single of int | `Range of int * int ]
+  ; chars : (int * int) option
+  }
+
+type report =
+  { loc : loc
+  ; message : message
+  ; related : (loc * message) list
+  }
+
+let dyn_of_severity =
+  let open Dyn.Encoder in
+  function
+  | Error -> constr "Error" []
+  | Warning { code; name } ->
+    constr "Warning" [ record [ ("code", int code); ("name", string name) ] ]
+
+let dyn_of_message =
+  let open Dyn.Encoder in
+  function
+  | Raw s -> constr "Raw" [ string s ]
+  | Structured { file_excerpt; message; severity } ->
+    constr "Structured"
+      [ record
+          [ ("file_excerpt", (option string) file_excerpt)
+          ; ("message", string message)
+          ; ("severity", dyn_of_severity severity)
+          ]
+      ]
+
+let dyn_of_loc { path; line; chars } =
+  let open Dyn.Encoder in
+  record
+    [ ("path", string path)
+    ; ( "line"
+      , match line with
+        | `Single i -> constr "Single" [ int i ]
+        | `Range (i, j) -> constr "Range" [ int i; int j ] )
+    ; ("chars", option (pair int int) chars)
+    ]
+
+let dyn_of_report { loc; message; related } =
+  let open Dyn.Encoder in
+  record
+    [ ("loc", dyn_of_loc loc)
+    ; ("message", dyn_of_message message)
+    ; ("related", list (pair dyn_of_loc dyn_of_message) related)
+    ]
+
+let re =
```

Comment:
Have you tried writing the parsing code as a lexer? This kind of code is usually easier to read when written as a lexer.

Reply:
That's how it was written initially. It did not work so well because there's no format for the error messages themselves. So we currently infer the messages to be whatever is between 2 error locations. That is something is naturally easier to write with `Re.split`. I could try doing it with the lexer again, but why fix what ain't broken? 😆 Hopefully this is something that the compiler will make obsolete once it adds json output (cc @octachron)

Reply:
Ok, that makes sense. Let's stick to the current code then.

Reply:
For the compiler output, I'd still be more keen on canonical S-expressions. A Json parser feels like it would bloat Dune, while a canonical S-expression printer is literally 5 lines of code.

Reply:
> For the compiler output, I'd still be more keen on canonical S-expressions. A Json parser feels like it would bloat Dune, while a canonical S-expression printer is literally 5 lines of code.

Agreed. I don't follow upstream development on this too closely, but it sounds like they already settled on json. You'd have to bring it up with them.

Reply:
I could certainly be convinced to add a sexp output (after all, the difference is small from the producer point of view). However, the idea that a mere json parser would bloat Dune sounds quite hyperbolic?

Reply:
It probably feels this way because the compiler doesn't need a json parser. FTR, Dune is a smaller software than the compiler (13M vs 16M for ocamlopt on my machine).